### ISSUE TITLE
[false-negative]QBA_QUESTIONABLE_BOOLEAN_ASSIGNMENT can not detect some complex case like

### ISSUE THREAD

**Issue by wuchiuwong** [spotbugs/spotbugs#1149]
**Affects SpotBugs Version:** 
spotbugs-4.0.0-RC2

**Rule:**
QBA_QUESTIONABLE_BOOLEAN_ASSIGNMENT

**Description:**
Unable to detect some complex case like: <code>if(foo = bar())</code> and <code>if (foo = bar == true)</code>
This rule is implemented in SpotBugs.jar\main\java\edu\umd\cs\findbugs\detect\QuestionableBooleanAssignment.java
The rule uses a complex method based on bytecode transfer to detect questionable boolean assignment, and there are special cases where this method is not applicable.

**Code Sample demonstrating the issue:**
```
example 1:
if(b = inKnowledge(g.directGoals)) {

example 2:
if (scanning = b == true)

example 3:
if (received = (barrier != null)) {
```

**Expected outcome:**
false-negative

**Running SpotBugs through:**
command line

**Comment by Xiaoven** [comment]
I don't think they are false negative. The description of this pattern is as [following](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#qba-method-assigns-boolean-literal-in-boolean-expression-qba-questionable-boolean-assignment)

> This method assigns a literal boolean value (true or false) to a boolean variable inside an if or while expression. Most probably this was supposed to be a boolean comparison using ==, not an assignment using =.

In your examples, it is hard to say that `if (scanning = b == true)` should be `if (scanning == b == true)`. Maybe the developer just want to judge `if ( b == true)` and assign the value of ` b == true` to the variable `scanning` which may be used in the following code.

Here is a real example I found:
```java
// https://github.com/TouK/sputnik-test/pull/3#discussion_r57453267
    private static void incorrectAssignmentInIfCondition() {
        boolean value = false;
        if (value = false) {
```

In this example, you can absolutely say it is questionable, because `if (value = false)` is equivalent to `if (false)` which makes no sense.
